### PR TITLE
ISSUE-373 Auto-complete contact integration test

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/373

### CODE REVIEW

**Comment by vttranlina** [comment]
TODO

- [ ] Wait https://github.com/linagora/tmail-backend/pull/392 merge 
- [ ]  Integration test for webadmin
- [ ] Integration test for AQMP

**Comment by vttranlina** [comment]
The integration test for Webadmin/AQMP will do in another tickets (https://github.com/linagora/tmail-backend/issues/397, https://github.com/linagora/tmail-backend/issues/398)

File: tmail-backend/integration-tests/contact-autocomplete/contact-autocomplete-integration-test-common/src/main/scala/com/linagora/tmail/integration/AutocompleteIntegrationContract.scala [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+             |			  "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+             |			  "list": [{
+             |         "id": "$${json-unit.ignore}",
+             |				  "firstname": "$${json-unit.ignore}",
+             |				  "surname": "$${json-unit.ignore}",
+             |				  "emailAddress":"${ANDRE.asMailAddress().asString()}"
+             |			  }],
+             |        "limit": 256
+             |		  },
+             |		  "c1"
```
Comment (quantranhong1999) [inline tmail-backend/integration-tests/contact-autocomplete/contact-autocomplete-integration-test-common/src/main/scala/com/linagora/tmail/integration/AutocompleteIntegrationContract.scala:129]: Please fix indent

File: tmail-backend/integration-tests/contact-autocomplete/memory-contact-autocomplete-integration-tests/src/test/resources/mailetcontainer.xml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                <onMailetException>ignore</onMailetException>
+            </mailet>
+            <mailet match="IsMarkedAsSpam" class="WithStorageDirective">
+                <targetFolderName>Spam</targetFolderName>
+            </mailet>
+            <mailet match="All" class="VacationMailet"/>
+            <mailet match="All" class="Sieve"/>
+            <mailet match="All" class="AddDeliveredToHeader"/>
+            <mailet match="All" class="org.apache.james.jmap.mailet.filter.JMAPFiltering"/>
+            <mailet match="All" class="com.linagora.tmail.mailets.ContactsCollection">
```
Comment (quantranhong1999) [inline tmail-backend/integration-tests/contact-autocomplete/memory-contact-autocomplete-integration-tests/src/test/resources/mailetcontainer.xml:106]: Should we use matcher SenderIs/SenderIsLocal ? 
rf: https://github.com/linagora/tmail-backend/pull/392#discussion_r845671045

**Review by quantranhong1999 — COMMENTED** [review]
Maybe 1 or 2 more tests: 
- Outsider sender (xyz@gmail.com) should not be index
- A case where auto-complete returns nothing when contact not related maybe

Overall good to me.

**Comment by vttranlina** [comment]
> * Outsider sender ([xyz@gmail.com](mailto:xyz@gmail.com)) should not be index

IMO That is the responsibility of mailet configuralation.  Here is mailet matcher (SenderIsLocal)

> A case where auto-complete returns nothing when contact not related maybe

I thought that, But I see we have `com.linagora.tmail.james.common.LinagoraContactAutocompleteMethodContract#contactAutocompleteShouldReturnEmptyListWhenNoContactFound` thus it is not really necessary?

File: tmail-backend/integration-tests/contact-autocomplete/contact-autocomplete-integration-test-common/src/main/scala/com/linagora/tmail/integration/AutocompleteIntegrationContract.scala [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+         |    "TMailContact/autocomplete",
+         |    {
+         |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |      "filter": {"text":"andre"}
+         |    },
+         |    "c1"]]
+         |}""".stripMargin
+
+    awaitAtMostTenSeconds.untilAsserted { () =>
+      val response: String = `given`
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply tmail-backend/integration-tests/contact-autocomplete/contact-autocomplete-integration-test-common/src/main/scala/com/linagora/tmail/integration/AutocompleteIntegrationContract.scala:100]: I do not understand. `response` is reused bellow

File: tmail-backend/integration-tests/contact-autocomplete/memory-contact-autocomplete-integration-tests/src/test/resources/mailetcontainer.xml (hunk 2 above)
Comment (vttranlina) [reply tmail-backend/integration-tests/contact-autocomplete/memory-contact-autocomplete-integration-tests/src/test/resources/mailetcontainer.xml:106]: I thought that first. But it is not clarified, which matcher we MUST choose. 
It just is optional, depending on the logic follow what we want. Then `All` will help us more focus on the auto-complete logic.

**Comment by vttranlina** [comment]
rebase master & fix comment

File: tmail-backend/integration-tests/contact-autocomplete/contact-autocomplete-integration-test-common/pom.xml [hunk 4]
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline tmail-backend/integration-tests/contact-autocomplete/contact-autocomplete-integration-test-common/pom.xml:1]: Personally I'm not fan of "yet 3 other maven modules". Can't we just put this in TMail JMAP integration tests for now? (Like we did for the encrypted mailbox?)

File: tmail-backend/integration-tests/contact-autocomplete/memory-contact-autocomplete-integration-tests/src/test/resources/mailetcontainer.xml (hunk 2 above)
Comment (chibenwa) [reply tmail-backend/integration-tests/contact-autocomplete/memory-contact-autocomplete-integration-tests/src/test/resources/mailetcontainer.xml:106]: Please use SenderIsLocal

**Review by Arsnael — COMMENTED** [review]
Read it, nothing to add

**Review by chibenwa — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Can you rebase ?

**Pull request merged**